cpp/src/decisiontree/batched-levelalgo/kernels.cuh

```diff
@@ -240,7 +240,9 @@ __global__ void nodeSplitKernel(IdxT max_depth, IdxT min_rows_per_node,
   auto range_start = node->start, range_len = node->count;
   auto isLeaf = leafBasedOnParams<DataT, IdxT>(
     node->depth, max_depth, min_rows_per_node, max_leaves, n_leaves, range_len);
-  if (isLeaf || splits[nid].best_metric_val <= min_impurity_decrease) {
+  if (isLeaf || splits[nid].best_metric_val <= min_impurity_decrease
```

Comment:
Interesting - our terminology here does not match scikit-learn's "min_samples_leaf" and min_samples_split (https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html). It looks like here you are cleanly implementing min_samples_leaf (don't create leaves with < x entries), but the docstring suggests it should work like min_samples_split:

"The minimum number of samples (rows) needed to split a node. If int then number of sample rows. If float the min_rows_per_sample*n_rows"

Maybe the docstring is wrong and this matches the old behavior?

Reply:
Yes, cuML RF only offers `min_rows_per_node`, whereas sklearn offers both `min_samples_leaf` and `min_samples_split`. The implication is that we use `min_rows_per_node` to constrain both internal and leaf nodes.

Reply:
this is a good catch, John! I think we should take this PR as an opportunity to enable it all the way from decision-tree-params in C++ to cuML-RF python API? What do you think @hcho3 ?

Reply:
@teju85 Do you mean that we should replace `min_rows_per_node` with `min_samples_leaf` / `min_samples_split`. That will involve an extensive modification. Do you think the risk is worthwhile?

Reply:
As per [here](https://github.com/rapidsai/cuml/blob/branch-0.17/cpp/include/cuml/tree/decisiontree.hpp#L51), `min_rows_per_node` is actually what sklearn calls it as `min_samples_leaf`. So, I was proposing to:
1. add a `min_samples_leaf` member variable into this struct [here](https://github.com/rapidsai/cuml/blob/branch-0.17/cpp/include/cuml/tree/decisiontree.hpp#L51)
2. update the set method for this struct [here](https://github.com/rapidsai/cuml/blob/branch-0.17/cpp/include/cuml/tree/decisiontree.hpp#L107) and update all its callers accordingly.
3. Expose this new variable in the cuML RF python API and update docs accordingly.
4. use this new variable in the below conditional instead of using `min_rows_per_node`.

Reply:
What do you guys think?

Reply:
@teju85 Good idea. Can we make the proposed change in another PR?

Reply:
I thought that this PR is basically addressing the issue of "only split the current node if the resulting leaf counts are greater than a certain threshold". So, to me, it sounds like this PR is ideal. Why do you think this deserves another PR? Or, if you think we should abandon this PR and file a new one, in light of the observation that John did, I'm fine with that too.

Reply:
I was afraid of making this PR very large. Besides, won't we also need to update the old backend to accommodate `min_samples_leaf`?

Reply:
I don't think we should fix old backend anymore. But would like to hear @JohnZed's thoughts as well on this one.

Proceeding with `min_rows_per_node` to check for split-decision is anyways incorrect as per John's findings. Thus, as I see we have 2 options: abandon this PR and file a new one with appropriate hyper-param added OR update this PR itself to do so.

Reply:
Got it. Let me give it a try and update this PR to add `min_samples_leaf` and `min_samples_split`.

Reply:
Maybe I'm wrong here, but as I understand it:
 * We have a variable called `min_rows_per_node` that is equivalent to sklearn's `min_samples_leaf` in both new and old backends (though previously not working in new backend)
 * We do NOT support an equivalent of `sklearn'`s `min_samples_split`
 * @teju85 's proposal would update the naming of this variable to be correct (in line with sklearn)

Assuming the above is correct, I'd suggest pushing in this PR as is (modulo any small changes requested) then changing the variable name to be min_samples_leaf everywhere. Since this is just a rename, I think supporting it in the old backend would be quite straightforward?

(For backward compatibility we can retain the parameter at the python layer but warn about deprecation and just use it to set min_samples_leaf)

Am I misunderstanding though?

Reply:
@JohnZed It might be good idea to support `min_samples_split`, to be on par with sklearn.

Reply:
No John. I meant to keep the `min_rows_per_node` name as-is (for now) but have this PR add support for `min_samples_split`, because that's what essentially we are solving with this PR.

Reply:
But if you want to also include renaming of `min_rows_per_node` to `min_samples_leaf` (to match sklearn's convention), I'm fine with reviewing that too.

Reply:
I think we're in the same ballpark of a solution... can we keep the PR as is but then have a follow-on PR that renames min_rows_per_node to min_samples_split? We should file an issue to add support for min_samples_split in the future, because that would improve compatibility, but I wouldn't want it to hold up this important improvement, especially as we get close to burndown time. How does that sound for next steps?

Reply:
We should not be using `min_rows_per_node` for the split-decision as we are currently doing in this PR. We should really be using `min_samples_split`.